### DISCUSSION TITLE
PDF viewer with fast reload

### DISCUSSION THREAD

**Discussion by Andrew15-5** [typst/typst#1981] [General]
I've used Okular for a long time with LaTeX. After a certain version, it stopped showing errors when reloading during PDF file generation, which is a big relief. Now it takes about 1 s to reload (tested with Typst). For LaTeX, I don't mind this extra 1 s, but since Typst compiles in ms/µs, I started to wonder if there is a PDF (pre-)viewer that reloads even faster than Okular.

I don't know where/how to find such a beast, so I was hoping that someone here could help/suggest something that they've used/know.

I use Linux and want a FOSS solution, but it can be anything, as long its refresh is sub-second. I'm a heavy dark theme user, so at least some kind of inversion of colors is desirable.

P.S. I use this live preview for Markdown: https://github.com/iamcco/markdown-preview.nvim. It is instant. I don't even have to save a file. I want same-level preview speed but for PDF files.

**TL;DR**: `mupdf` has no flickering whatsoever during file reloading (tested for a few minutes), and it feels like the PDF viewer with the fastest file reloading time. `evince` is slightly slower with a lot of flickering (especially on each keystroke reload). `zathura` has less flickering than `evince` but it crushed several times during testing — not good. `typst-preview` is in beta and have many small/big issues, but is very fast and has small/no flickering (could be wrong). `okular` is very slow for me (+1 s), but others have measured 200 ms, but it is better (best) overall FOSS PDF viewer.

### `evince`

I've tried GNOME's `evince` and it is indeed refreshes way under 1 s (amazing!). With screen capture, I've measured refresh time around 100–200 ms. For Typst documents, `evince` is a great alternative (I don't need anything from it except color inversion and quick refresh, unlike Okular for everything else).

---

I've tried `typst-preview` but I so far it didn't work (https://github.com/Enter-tainer/typst-preview/issues/93), which is a big shame.

## Update 2

I've tried `zathura`, `mupdf` and `typst-preview` also now works.

### `zathura`
`zathura` is a document viewer with vim navigation (if vim was a document viewer). It supports inverting colors and status bar can be hidden (I use Neovim all the time). When page is reloading, it quickly shows "Loading" in the center, and it has small text flashing and "jitter" during reloading. Unfortunately, it did crash at least 3 times (probably memory issue in C), so maybe I will open an issue (https://git.pwmt.org/pwmt/zathura). It flashes less than `evince`.

### `mupdf`
`mupdf` is also looks minimalistic as `zathura` and also have some keyboard shortcuts (also can invert colors). I did the `while inotifywait -e modify file.pdf ; do pkill -HUP mupdf; done` so that it will reload the file on every modification (generation) of the PDF file. And it just doesn't flash AT ALL! It's kinda perfect, although I've tested it only for a few minutes. I've (finally) enabled auto save on each keystroke in `INSERT` mode and when I type any letter, the only thing that happens in the `mupdf` window, is a new letter appears. That's it. Although it's a shame that I need a second process that will send signals (`HUP`) to `mupdf` on every file change, but the overall experience is unbelievable.

### Missing `okular`'s different methods of color inversion
One downside as compared to `okular` is missing "smart" color inversion:

![image](https://github.com/typst/typst/assets/37143421/271def16-8fd3-4e18-bf68-728d7de81d58)

It can help when standard "negative" (always) makes images ugly (AF). But how images look isn't a big issue for me.

### `typst-preview`
Finally, `typst-preview`. At the time of writing (2023-08-23Z) it has problems with manual compilation, but release binaries just work. Right now, you have to add `--open-in-browser` for it to work. After that is look kinda good, but I need to tweak "Dark Reader" extension to be able to see white text on black. One noticeable problem is zoom. A4 format doesn't fit whole, so I zoom out a bit. After modifying the source file, it will reset to default zoom, but browser's zoom setting is still present. So basically I can't zoom out and lock it like that. I also noticed degrading performance with a bunch of `#lorem(100)`. It has some optimization options, but it is intended to be used in VS Code, so `partial-rendering` (I think) doesn't work, "click to jump" obviously doesn't work. Sometimes it can just show an empty page because of something (I had a bunch of lorems and was adding 1 letter at a time). So, it could be a great alternative, but it's in the beta, so it has a lot of time to improve.

## Update 3

Unfortunately, after a long time of daily driving the `mupdf` I can tell that it too can sometime crash (as it's the case with `zathura`). I only tested it with `typst-lsp`'s built-in Typst compiler. It also can sometime show a full window thick red cross. To fix it, I need to just undo-redo the last file change in a text editor.

The good news is that `typst-preview` is becoming less buggy, and now I can actually build it from source (which wasn't the case when I wrote the last update) instead of waiting for a new release binary. It still has a few bugs (https://github.com/Enter-tainer/typst-preview/issues/130, https://github.com/Enter-tainer/typst-preview/issues/131, https://github.com/Enter-tainer/typst-preview/issues/132) which are a dealbreaker for me, but after they are all resolved, I think I can switch to it and test it out.

**Comment by alerque** [comment]
First, what platform are you on?

Second, how are you managing your file writes and re-renders?

Okular on my 7 year old PC laptop (testing on a 24 page text document) re-renders new PDF file contents in something less than 200ms after file writes complete. For me `zathura` is a bit faster reading files coming in the 70ms range, but it also flashes a bit more when redrawing a page. Even the clumsy Gnome builtin viewer Evince takes way less than a second.

I suspect the bigger issues is going to be file writes and the time it takes for your editor to flush the buffer to disk and trigger a file saved event. Many markdown previewers work by catching the event inside the editor itself and responding incrementally to the changes in-bufffer.

**Reply by Andrew15-5** [reply]
Since when Linux isn't a platform? Ok, Debian based OS for the near future.

**Reply by Andrew15-5** [reply]
I used evince at first, but I've noticed that it can't render some objects in presentation PDF properly, so I moved to okular. I'll try it again.

**Reply by alerque** [reply]
Linux is of course a platform, I just missed that in your post even after looking for it.

Evince is not my suggestion since it blinks like crazy when redrawing *and* is the slowest. I only mentioned it because for me on old hardware it comes in way less than a second, which makes me think you have some other problem with your pipeline.

Give Zathura a shot.

Alsa `mupdf` might be worth looking into. It doesn't have an auto-refresh but you can make it refresh externally by sending the process a signal. I found a long time ago when I was playing around with PDF previews that mupdf was the least-blinky way get an updating preview window. You'd need to combine the tooling with `entr` or similar to get it refreshing—something like `typst watch & ; mupdf foo.pdf & ; echo foo.pdf | entr pkill -HUP mupdf`

**Reply by Andrew15-5** [reply]
I've updated the top message (you can take a look). You've helped a lot, thank you!

**Comment by PgBiel** [comment]
If you're just looking for a preview, you could try using tools based on PNG export instead, such as [the Typst Preview VSCode Extension](https://github.com/Enter-tainer/typst-preview) (note: unofficial).

**Reply by Andrew15-5** [reply]
Wow, this sounds very promising, I'll check it out. It's based on _SVG_ incremental builds/renders.

**Reply by Andrew15-5** [reply]
I've tried working version. It still looks very promising, but has a lot of issues and in beta. So it's not production-ready, i.e. I can't comfortably use it. Not yet. You can check out my updated message for much more information.
